src/nmodl/nocpout.cpp

```diff
@@ -3137,3 +3163,148 @@ Symbol* breakpoint_current(Symbol* s) {
+// Need to recursively look through func_calls but watch out for loops.
+// If there are no RANGE then VERBATIM is ok and set all need_setdata to false.
+
+static bool check_func(Symbol* s);  // recursive
+
+void func_needs_setdata() {
+    if (suffix[0] == '\0') {
+        return;
+    }
+    for (auto& f: funcs) {
+        f.second.is_being_looked_at = false;
+    }
+
+    // if there are no RANGE then set all need_setdata to false.
+    bool norange{true};
+    Item* q;
+    int i;
+    SYMLISTITER {
+        Symbol* s = SYM(q);
+        if (s->type == NAME && s->nrntype & (NRNRANGE | NRNPOINTER)) {
+            norange = false;
+            break;
+        }
+    }
+    if (norange) {
+        for (auto& f: funcs) {
+            f.second.need_setdata = false;
+        }
+    }
+
+    for (auto& f: funcs) {
+        check_func(f.first);
+    }
+    for (auto& f: funcs) {  // update the hocfunc item if need_setdata
+        auto& q = f.second.q;
+        if (q && f.second.need_setdata) {
+            // error if not valid id
+            Symbol* s = f.first;
+            Sprintf(buf,
+                    "\n"
+                    "  if(!_prop_id) {\n"
+                    "    hoc_execerror(\""
+                    "No data for %s_%s. Requires prior call to setdata_%s"
+                    " and that the specified mechanism instance still be in existence.\","
+                    " NULL);\n",
+                    s->name,
+                    mechname,
+                    mechname);
+            insertstr(q, buf);
+            if (vectorize) {
+                insertstr(q,
+                          " }\n"
+                          "  Prop* _local_prop = _extcall_prop;\n");
+            } else {
+                // ensure current instance matches _extcall_prop
+                insertstr(q,
+                          " } else {\n"
+                          "    _setdata(_extcall_prop);\n"
+                          "  }\n");
+            }
+
+        } else if (q) {
+            if (vectorize) {
+                // if id not valid then _local_prop must be nullptr
+                // because of later _ppvar = _local_prop ? ...
+                insertstr(q, "\n  Prop* _local_prop = _prop_id ? _extcall_prop : nullptr;\n");
+            }
+        }
+    }
+}
+
+static bool check_func(Symbol* s) {  // recursive
+    if (funcs.count(s) == 0) {
+        return false;
+    }
+    Info& i = funcs[s];
+    if (i.need_setdata) {
+        return true;
+    }
+    if (i.is_being_looked_at) {
+        return false;
+    }
+    i.is_being_looked_at = true;
+    for (auto& s1: i.func_calls) {
+        if (check_func(s1)) {
+            i.need_setdata = true;
+            return true;
+        }
+    }
+    return false;
+}
+
+// If the function needs setdata, then q can be changed to
+// perform the check on _extcall_prop
+// Not called for POINT_PROCESS functions.
+void hocfunc_setdata_item(Symbol* s, Item* q) {
+    auto& i = funcs[s];
+    i.q = q;
```

Comment:
I guess/assume no check is needed to see if `funcs` contain `s`.

Reply:
I was relying on C++ to raise an error as that would infer a bug in the nmodl implementation. Good catch! I think that will happen with FUNCTION_TABLE (which does not use RANGE variables). I'll add that to my test and think about how to deal with that case.